Remove userAgentAppearanceStyle

anttijk opened the pull request:
#### 93f03ffffb803aebc218be3fb210db9672bc99a2
<pre>
Remove userAgentAppearanceStyle
<a href="https://bugs.webkit.org/show_bug.cgi?id=286288">https://bugs.webkit.org/show_bug.cgi?id=286288</a>
<a href="https://rdar.apple.com/143305321">rdar://143305321</a>

Reviewed by Aditya Keerthi.

We now track in style builder if author mutated border or background properties that disable
native control styling. Checking for this is the only reason we generate and cache this style.

The userAgentAppearanceStyle is already unused and maintained unnecessarily when &quot;devolved widget&quot;
feature is enabled (which it is by default). This patch aligns the legacy case with the standard too
by having any author mutation to border and background style disable the native appearance
(rather than requiring it to bet set to something that differs from UA style). The legacy list of
affected controls is maintained.

* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::hasAppearanceForElementTypeFromUAStyle):
(WebCore::RenderTheme::adjustStyle):
(WebCore::RenderTheme::isControlStyled const):
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.cpp:
(WebCore::RenderThemeAdwaita::isControlStyled const):
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.h:
* Source/WebCore/rendering/ios/RenderThemeIOS.h:
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::isControlStyled const):
* Source/WebCore/rendering/mac/RenderThemeMac.h:
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
(WebCore::RenderThemeMac::isControlStyled const):
* Source/WebCore/style/MatchResultCache.cpp:
(WebCore::Style::copy):
* Source/WebCore/style/MatchedDeclarationsCache.cpp:
(WebCore::Style::MatchedDeclarationsCache::add):
* Source/WebCore/style/MatchedDeclarationsCache.h:
* Source/WebCore/style/ResolvedStyle.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
(WebCore::Style::Adjuster::adjustAnimatedStyle const):
(WebCore::Style::Adjuster::adjustThemeStyle const):
* Source/WebCore/style/StyleAdjuster.h:
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyProperty):
* Source/WebCore/style/StyleBuilderState.cpp:
(WebCore::Style::BuilderState::disableNativeAppearanceIfNeeded):

Factor into a function in BuilderState.
Set the nativeAppearanceDisabled bit also in the legacy hasAppearanceForElementTypeFromUAStyle case.

* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::State::takeStyle):
(WebCore::Style::Resolver::unadjustedStyleForElement):
(WebCore::Style::Resolver::styleForElement):
(WebCore::Style::Resolver::unadjustedStyleForCachedMatchResult):
(WebCore::Style::Resolver::styleForKeyframe):
(WebCore::Style::Resolver::styleForPseudoElement):
(WebCore::Style::Resolver::applyMatchedProperties):
(WebCore::Style::Resolver::State::takeUserAgentAppearanceStyle): Deleted.
(WebCore::Style::Resolver::State::userAgentAppearanceStyle const): Deleted.
(WebCore::Style::Resolver::State::setUserAgentAppearanceStyle): Deleted.
(WebCore::Style::elementTypeHasAppearanceFromUAStyle): Deleted.
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::styleForStyleable):
(WebCore::Style::TreeResolver::resolveAgainInDifferentContext):

Canonical link: <a href="https://commits.webkit.org/294554@main">https://commits.webkit.org/294554@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c53d61bb359350bf3c6ab7dee003deb0a2223de9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102297 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21966 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12280 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107458 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52933 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104336 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22274 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30472 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/77840 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34827 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105303 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17243 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92344 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58178 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/17077 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10372 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52291 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86902 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10442 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109832 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29429 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/21696 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/86820 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29793 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/88543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86409 "Found 1 new API test failure: /WebKitGTK/TestCookieManager:/webkit/WebKitCookieManager/persistent-storage (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21973 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31219 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/8938 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/23671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29357 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34652 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29168 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32491 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30727 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/rendering/mac/RenderThemeMac.mm:
```diff
@@ -695,11 +695,11 @@ static Color activeButtonTextColor()
     auto appearance = style.usedAppearance();
     if (appearance == StyleAppearance::TextField || appearance == StyleAppearance::TextArea || appearance == StyleAppearance::SearchField || appearance == StyleAppearance::Listbox)
-        return style.border() != userAgentStyle.border();
+        return style.nativeAppearanceDisabled();
```
This is technically a behavior change for apps linked before `SDKAlignedBehavior::DevolvableWidgets`.

pxlcoder replied:
After more consideration, this is probably okay. The implication would be that apps linked before `SDKAlignedBehavior::DevolvableWidgets` would no longer be able to customize the background color of their text fields, text areas, and <select multiple> without also losing native appearance.

However, the control would simply devolve and their specified background color would be honored so I would not expect there to be any contrast/legibility issues. Which leaves a minor cosmetic change without impact to functionality.

pxlcoder commented on Source/WebCore/style/StyleBuilder.cpp:
```diff
@@ -401,7 +401,7 @@ void Builder::applyProperty(CSSPropertyID id, CSSValue& value, SelectorChecker::
 
     if (!isUnset) {
-        if (cascadeLevel == CascadeLevel::Author && m_state.element()->isDevolvableWidget() && CSSProperty::disablesNativeAppearance(id) && m_state.applyPropertyToRegularStyle())
+        if (cascadeLevel != CascadeLevel::UserAgent && CSSProperty::disablesNativeAppearance(id) && m_state.applyPropertyToRegularStyle())
```
Note that the spec explicitly mentions the author origin: https://drafts.csswg.org/css-ui/#appearance-disabling-properties

Though this does seem like a case where the spec should probably be updated, I don't think it makes sense to exclude the User Origin from this behavior, as we do right now.

pxlcoder replied:
I filed https://github.com/w3c/csswg-drafts/issues/11555.

pxlcoder replied:
Given the resolution of that issue, it seems we should leave as `cascadeLevel == CascadeLevel::Author`.

anttijk replied:
reverted this change

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -212,7 +212,20 @@ static bool shouldCheckLegacyStylesForNativeAppearance(const Element* element)
+        || localName == HTMLNames::progressTag
+        || localName == HTMLNames::selectTag
+        || localName == HTMLNames::meterTag
+        || (element.isInUserAgentShadowTree() && element.userAgentPart() == UserAgentParts::webkitListButton());
```
@lilyspiniolas is also adding `appearance` (internal-only) to parts of the color input, so this will need to be updated depending on the order of landing.

lilyspiniolas replied:
Yes, there is now `UserAgentParts::webkitColorSwatchWrapper()`, `UserAgentParts::webkitColorSwatch()`, and `UserAgentParts::internalColorSwatchOverlay()`

anttijk replied:
I thought we'd want to keep the legacy (linkedBefore) behavior unchanged and not add new author styleable controls there?

pxlcoder replied:
We don't ever need to add `internalColorSwatchOverlay` since it's not web-exposed, but the swatch wrapper and color swatch (which have long existed) should probably be added for correctness.

Though reading this PR again, I see this change preserves the existing behavior for those elements, so that is unrelated and can be done separately.

anttijk committed:
Remove userAgentAppearanceStyle
https://bugs.webkit.org/show_bug.cgi?id=286288
rdar://143305321

Reviewed by Aditya Keerthi.

We now track in style builder if author mutated border or background properties that disable
native control styling. Checking for this is the only reason we generate and cache this style.

The userAgentAppearanceStyle is already unused and maintained unnecessarily when "devolved widget"
feature is enabled (which it is by default). This patch aligns the legacy case with the standard too
by having any author mutation to border and background style disable the native appearance
(rather than requiring it to bet set to something that differs from UA style). The legacy list of
affected controls is maintained.

* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::hasAppearanceForElementTypeFromUAStyle):
(WebCore::RenderTheme::adjustStyle):
(WebCore::RenderTheme::isControlStyled const):
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.cpp:
(WebCore::RenderThemeAdwaita::isControlStyled const):
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.h:
* Source/WebCore/rendering/ios/RenderThemeIOS.h:
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::isControlStyled const):
* Source/WebCore/rendering/mac/RenderThemeMac.h:
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
(WebCore::RenderThemeMac::isControlStyled const):
* Source/WebCore/style/MatchResultCache.cpp:
(WebCore::Style::copy):
* Source/WebCore/style/MatchedDeclarationsCache.cpp:
(WebCore::Style::MatchedDeclarationsCache::add):
* Source/WebCore/style/MatchedDeclarationsCache.h:
* Source/WebCore/style/ResolvedStyle.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
(WebCore::Style::Adjuster::adjustAnimatedStyle const):
(WebCore::Style::Adjuster::adjustThemeStyle const):
* Source/WebCore/style/StyleAdjuster.h:
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyProperty):
* Source/WebCore/style/StyleBuilderState.cpp:
(WebCore::Style::BuilderState::disableNativeAppearanceIfNeeded):

Factor into a function in BuilderState.
Set the nativeAppearanceDisabled bit also in the legacy hasAppearanceForElementTypeFromUAStyle case.

* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::State::takeStyle):
(WebCore::Style::Resolver::unadjustedStyleForElement):
(WebCore::Style::Resolver::styleForElement):
(WebCore::Style::Resolver::unadjustedStyleForCachedMatchResult):
(WebCore::Style::Resolver::styleForKeyframe):
(WebCore::Style::Resolver::styleForPseudoElement):
(WebCore::Style::Resolver::applyMatchedProperties):
(WebCore::Style::Resolver::State::takeUserAgentAppearanceStyle): Deleted.
(WebCore::Style::Resolver::State::userAgentAppearanceStyle const): Deleted.
(WebCore::Style::Resolver::State::setUserAgentAppearanceStyle): Deleted.
(WebCore::Style::elementTypeHasAppearanceFromUAStyle): Deleted.
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::styleForStyleable):
(WebCore::Style::TreeResolver::resolveAgainInDifferentContext):

Canonical link: https://commits.webkit.org/294554@main
